File: trees/c++/08_1.cpp
#include<iostream>

struct node
{
    int data;
    node* left, *right;
};

node* get_new_node(int data)
{
    node* temp = new node();
    temp->data = data;
    temp->left = temp->right = NULL;
    return temp;
}

node* insert(node* root, int data)
{
    if (!root)
        root = get_new_node(data);
    else if (data < root->data)
        root->left = insert(root->left, data);
    else
        root->right = insert(root->right, data);
    return root;
}

void inorder(node* root)
{
    if (root)
    {
        inorder(root->left);
        std::cout<<root->data<<" ";
        inorder(root->right);
    }
}

node* merge_two_node(node* first, node* second)
{
    /*
    if (!first && !second)
        return first;
    if (!first)
        return second;
    if (!second)
        return first;
    */

    while (first->right)
        first = first->right;
    while (second->left)
        second = second->left;

    first->right = second;
    second->left = first;

    return first;
    /*
    // to return the pointer to first node of first dll
    while (first->left->left)
        first = first->left;
    return first;
    */
}

node* make_dll(node* root)
{
    if (root)
    {
        node* l_node, *r_node;
        l_node = make_dll(root->left);
        r_node = make_dll(root->right);
        root = merge_two_node(merge_two_node(l_node, root), r_node);
        return root;
    }
}

void iterate_ll(node* head)
{
    node* temp = head;

    while (temp)
    {
        std::cout<<temp->data<<" ";
        temp = temp->right;
    }
}

int main()
{
    node* root = NULL;

    root = insert(root, 100);
    root = insert(root, 50);
    root = insert(root, 70);
    root = insert(root, 80);
    root = insert(root, 110);

    //inorder(root);
    std::cout<<std::endl;

    root = make_dll(root);

    iterate_ll(root);
    std::cout<<std::endl;
    return 0;
}

File: trees/python/08_1.py
import bst


# the solution in my view can be made by focusing on the fact that a single
# leaf node can be considered as a DLL, then using it as basic building block
# for constructing the lest part of the problem
def constructDLL(root):
    # I want to first get the DLL from LST
    if root.left:
        lstDLL = constructDLL(root.left)
        # now we wanna reach to the tail of the DLL, so we go to extreme right
        # in other words we can also say that we're finding inorder predecessor
        # (but only if we consider the tree as BST)
        while lstDLL.right:
            lstDLL = lstDLL.right
        # also we need to store the head as the right pointer of the lstDLL, so
        lstDLL.right = root  # root is the head, so we make it point to root
        root.left = lstDLL  # since, this is DLL, so we need both pointer to
        # point at each other

    if root.right:
        rstDLL = constructDLL(root.right)
        # in this case we need to go the extreme left
        while rstDLL.left:
            rstDLL = rstDLL.left
        # this time we don;t have to store the root (i.e. head ptr), rather we
        # need to store the tail into the left pointer which is storing head
        rstDLL.left = root  # root is the tail in this case
        root.right = rstDLL

    return root


def printDLL(head):
    temp = head
    while temp:
        print temp.data,
        temp = temp.right


if __name__ == '__main__':
    root = None
    BSTree = bst.BST()
    inp = input('enter the elements of tree separated by ,\n')
    for i in inp:
        root = BSTree.insert(root, i)

    root = constructDLL(root)

    while root.left is not None:
        root = root.left

    printDLL(root)
